Refactor groups service's `load` method

robertknight opened the pull request:
The `load` method in the groups service that fetches groups from the API, adds them to the store and focuses the appropriate group, was getting difficult to follow.

This PR is a series of incremental refactorings of this method to hopefully make it easier to follow. I've also added some high-level comments to clarify the main steps involved.

A commit-by-commit review will probably be easier than looking at the diff all at once.

robertknight commented:
I've noticed an issue with the tests often producing an "unhandled promise rejection" error but still passing. I'm going to look into this but I don't currently think it is related to the changes here.

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+        // don't already have it.
+
+        // If there is a direct-linked group, add it to the featured groups list.
+        let allFeaturedGroups =
```
I realize you inherited it, but this conditional/ternary is a bit of a mouthful. Might help to extend the comment above to indicate that it (the direct-linked group) will be added to the featured-groups list _if it's not there already_. (I think part of what threw me is the naming of `selectedGroup` here, though I understand _why_ it's called that).

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+        let allFeaturedGroups =
+          selectedGroup !== null &&
+          !featuredGroups.some(g => g.id === selectedGroup.id)
+            ? featuredGroups.concat([selectedGroup])
```
If I'm reading this correctly, the direct-link annotation's group is at the end of the featured-groups list if it wasn't in the featured groups list to start with, but could be anywhere within it if it was already there (unless sorting happens later)? Is this intentional? Could it cause an inconsistency?

(edited: Looks like maybe the `combineGroups` stuff later takes care of sorting?)

robertknight replied:
From a quick scan, it looks like groups are sorted by type in `groups.all` and later sorted by organization in `groupsByOrganization` when presenting groups. I can't see anything which sorts them in lexicographic order here. Up until now it looks like the client relied on the API response to sort groups by name and then by ID. So it looks like this could potentially cause the direct-linked group to be in non-alphabetical order in the store.

Since there is typically only one direct-linked group and it appears in its own "Currently viewing" section in the groups menu UI however, I don't think that will actually matter.

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+            ? featuredGroups.concat([selectedGroup])
+            : featuredGroups;
+
+        // If there's a selected annotation it may require an extra api call
```
Might I request that we capitalize "API" in comments?

lyzadanger replied:
What's the difference between a "selected annotation" and a "direct-linked annotation"? Can an annotation be selected in more ways than directly linking to it?

robertknight replied:
> Might I request that we capitalize "API" in comments?

Done.

> What's the difference between a "selected annotation" and a "direct-linked annotation"? 

There is none as far as I am aware. I've changed this for consistency.

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+
+        // If there's a selected annotation it may require an extra api call
+        // to fetch its group.
+        if (selectedAnn) {
```
I notice inconsistency in naming (`selectedAnn` vs. `directLinkedAnnotationGroupId`). Is there a historical reason for this? Is it possible to be more consistent (i.e. spell out _annotation_ versus abbreviation)?

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+          // the filterGroups method.
+          directLinkedAnnotationGroupId = selectedAnn.group;
+
+          const selectedAnnGroup = myGroups
```
I'm having trouble with the "selected" vs. "directLinked" terminology again here...

`selectedAnnGroup` is a Boolean, right (return value of `Array.prototype.some`)? The naming here is confusing as it implies an object/thing.

I feel like this line could use a comment along the lines of "see if the selected annotation's group is already in our known list of groups" (assuming I'm even parsing that correctly :) )

robertknight replied:
I agree. I'm not sure why I didn't see this as an issue when I reviewed the original code. I've pushed a change to use the term "direct-linked" consistently.

lyzadanger commented on src/sidebar/services/groups.js:
```diff
@@ -225,75 +234,68 @@ function groups(
+
+          // If the direct-linked annotation's group has not already been fetched,
+          // fetch it.
+          if (!selectedAnnGroup) {
```
I find this Promise block challenging to parse. Maybe renaming the Boolean that indicates whether a group-fetch is needed and a slight restructuring something like the following?

```
if(weNeedToFetchTheGroup) { // obviously not a real name here

  const groupApi = fetchGroup({ /* ... */ });

  return groupApi.then(group => {
    if(group) {
      return allFeaturedGroups.concat(group);
    }
    return allFeaturedGroups;
  });
}
```

Specifically, I think the reassignment to `initialFeaturedGroups` feels confusing.

robertknight commented:
Thanks for the feedback Lyza. I've addressed the "selected" vs "direct-linked" naming confusion and also adjusted comments. There is probably more that can be done here but I'm going to merge at this point.
